specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1082 @@
+
+In order for a new checkpoint to be accepted for commitment in `SCA`, the source of the message `CommitChildCheckpoint()` needs to be the address of the subnet actor of the corresponding subnet; the subnet needs to be in an `Active` state (i.e. its collateral is over `CollateralThreshold`); the epoch of the checkpoint should be a multiple of `CheckPeriod` and larger that that of the previous checkpoint; and the checkpoint must point to the `CID` of the previous checkpoint committed by the subnet. 
+
+When spawned, subnets are allowed to configure the `CheckPeriod` that better suits their needs and the specifics of its consensus algorithm.
+
+### Checkpoint Commitment
+As an example, lets consider a checkpoint for subnet `/root/f0100/f0200`. Every `CheckPeriod` (in terms of subnet block time), validators access the checkpoint template that needs to be signed and populated by calling the `CheckpointTemplate()` state accesor from `SCA` in `/root/f0100/f0200`. Once signed and populated, checkpoints from `/root/f0100/f0200` are submitted to the `SA` with ID `f0200` of the subnet network `/root/f0100` by sending a message to `SubmitCheckpoint()`. After performing the corresponding checks and waiting for the commitment conditions (i.e. `2/3` of the validators sending a signed checkpoint in its reference implementation), this actor then triggers a message to `/root/f0100` `SCA`'s `CommitChildCheckpoint()` function to commit the checkpoint.
+
+When the checkpoint is committed, `SCA` in `/root/f0100` is responsible for aggregating the checkpoint from `/root/f0100/f0200` with those of other children of `/root/f0100`, and for generating a new checkpoint for `/root/f0100` that is then propagated to its parent chain, `/root`. The committment of checkpoints also triggers the execution and propagation of cross-net messages. As checkpoints flow up the chain, the `SCA` of each chain picks up these checkpoints and inspects them to propagate potential state changes triggered by messages included in the cross-net messages that have the `SCA`'s subnet as a destination subnet (see [Cross-net messages](#Cross-net-messages "Cross-net messages")).
+
+Every checkpoint for a subnet points to the previous checkpoint being committed to ensure their integrity.
+
+> Checkpoint template state accessor in SCA
+```go
+// Returns the checkpoint template to be populated for the current
+// checkpoint window. The current checkpoint window is computed by 
+// (currEpoch / CheckPeriod) * CheckPeriod. This determines the checkpoint
+// that needs to be populated and signed for the current window.
+CheckpointTemplate()
+```
```

Comment:
It's not entirely clear what "state accessor" is, whether it's something special or just a read-only method on the SCA state. I thought it's that, but this function definition doesn't have a return value. Not sure if it's just missing, or is it supposed to send back a message (although I don't think that would be a helpful distinction in this case).

Reply:
Really well spotted. It is a read-only function and it should return a checkpoint. I just fixed it.